Simplify backports

dieggsy opened the pull request:
Addresses #258. As I mentioned there, we could also do without `if-let*` and `when-let*`, but those are used much more extensively, so waiting for your opinion on that.

raxod502 commented on straight.el:
```diff
@@ -3970,10 +3909,10 @@ This function also adds the recipe repository to
+  (straight--alist-set
+   (car melpa-style-recipe)
+   (cdr melpa-style-recipe)
+   (alist-get straight-current-profile straight-recipe-overrides)))
```
`straight--alist-set` does not actually mutate the original alist. You have to re-assign it to the original place. That will probably be a little inconvenient here because of the nested `alist-get` usage, but not too bad.

raxod502 replied:
Also, if the keys are symbols (which I believe they are in this case), it's more efficient to pass `'symbol` as the fourth argument to `straight--alist-set`, so that `eq` is used for comparison.

raxod502 commented on straight.el:
```diff
@@ -4627,7 +4566,10 @@ is loaded, according to the value of
+         (unless (alist-get :straight use-package-defaults)
+           (setq use-package-defaults
+                 (delq (assoc :straight use-package-defaults)
+                       use-package-defaults)))))))
```
I think it's better to use `assq-delete-all` for this.

raxod502 commented on straight.el:
```diff
@@ -4670,8 +4612,10 @@ is loaded, according to the value of
+           (straight--alist-set
+            :straight
+            '('(t) straight-use-package-by-default)
+            use-package-defaults)))))))
```
Same as above, this needs an extra `setq`.

raxod502 replied:
We can also use `'symbol` here.

raxod502 commented:
This doesn't have to be addressed in this pull request, but I think the compatibility definitions of `alist-get`, `if-let*`, and `when-let*` could be simplified because we don't use many of the features, including the `REMOVE` argument to `alist-get`, the ability of `if-let*` and `when-let*` to take nonstandard binding vectors, etc.

Plus it would probably be smarter to condition on `fboundp` rather than checking the version of Emacs.

dieggsy commented:
I've rebased the appropriate commit with your suggestions (the `straight-override-recipe` one is a bit convoluted, not sure if I got it right) and added an alist-get simplification. Haven't looked in depth into the `internal--*` functions and `if-get*` logic yet though.

raxod502 commented on straight.el:
```diff
@@ -208,14 +147,13 @@ VARLIST is the same as in `if-let*'."
 
 This is a generalized variable suitable for use with `setf'.
 When using it to set a value, optional argument REMOVE non-nil
 means to remove KEY from ALIST if the new value is `eql' to DEFAULT."
```
Probably remove reference to `REMOVE` from the docstring too.

dieggsy replied:
Good point. Removing the setf bit as well, since we removed the gv- definition

raxod502 commented on straight.el:
```diff
@@ -4627,7 +4571,9 @@ is loaded, according to the value of
        (when (and (boundp 'use-package-defaults)
                   (listp use-package-defaults))
-         (setf (alist-get :straight use-package-defaults nil 'remove) nil)))))
+         (unless (alist-get :straight use-package-defaults)
```
Why do we have an `unless` here? It seems like it would cause the `setq` below to never be executed.

dieggsy replied:
Because of this, I think:
```
When using it to set a value, optional argument REMOVE non-nil
means to remove KEY from ALIST if the new value is eql to DEFAULT.
```
So as far as I understand the previous code was simply removing that key if the value of `:straight` was `nil`. The unless makes sure the value isn't something other than nil before removing it.

raxod502 replied:
No, I think "new value" in `if the new value is eql to DEFAULT` means the value that's provided to `setf`, not the value in the alist. Certainly I did not intend to condition removal on the value being nil (since it was never supposed to be nil).

dieggsy replied:
Ah, ok, I think I misunderstood the docs. So to be clear, you basically whant to always remove the `:straight` key from `use-package-defaults` within the enclosing `when` statement? So removing the `unless` should be enough.

raxod502 replied:
Yes, your assessment is correct.

dieggsy replied:
Ok, I've made the change.

raxod502 commented:
Thanks!
